h/views/api_auth.py

```diff
@@ -85,6 +112,22 @@ def _authorized_response(self):
             client_id = self.request.params.get('client_id')
             raise RuntimeError('created authorisation code for client "{}" but got no redirect location'.format(client_id))
 
+    def _render_web_message_response(self, redirect_uri):
```

Comment:
In a similar vein to the question above about wrapping and unwrapping HTTP requests, this method feels like it’s taking a bunch of information that’s been wrapped up into a URL, unwrapping it again to extract the `code`, `state` and `origin`, then using those to render the template. Is this something that’s forced upon us by the way `oauthlib` does things, or is there a way to pass these values around directly and only build them together into URLs at the point we need them?